## Gather

In [205]:
import pandas as pd

import numpy as np

In [206]:
patients = pd.read_csv('patients.csv')
treatments = pd.read_csv('treatments.csv')
adverse_reactions = pd.read_csv('adverse_reactions.csv')

## Assess

In [207]:
patients

patient_id assigned_sex given_name      surname  \
0             1       female        Zoe      Wellish   
1             2       female     Pamela         Hill   
2             3         male        Jae       Debord   
3             4         male       Liêm         Phan   
4             5         male        Tim      Neudorf   
..          ...          ...        ...          ...   
498         499         male    Mustafa    Lindström   
499         500         male      Ruman      Bisliev   
500         501       female      Jinke    de Keizer   
501         502       female    Chidalu  Onyekaozulu   
502         503         male        Pat      Gersten   

                       address               city       state  zip_code  \
0         576 Brown Bear Drive  Rancho California  California   92390.0   
1    2370 University Hill Road          Armstrong    Illinois   61812.0   
2        1493 Poling Farm Road               York    Nebraska   68467.0   
3          2335 Webster Street         Woodbridge          NJ    7095.0   
4         1428 Turkey Pen Lane             Dothan          AL   36303.0   
..                         ...                ...         ...       ...   
498        2530 Victoria Court       Milton Mills          ME    3852.0   
499   494 Clarksburg Park Road             Sedona          AZ   86341.0   
500          649 Nutter Street      Overland Park          MO   64110.0   
501   3652 Boone Crockett Lane            Seattle          WA   98109.0   
502          2778 North Avenue               Burr    Nebraska   68324.0   

           country                                          contact  \
0    United States             951-719-9170ZoeWellish@superrito.com   
1    United States            PamelaSHill@cuvox.de+1 (217) 569-3204   
2    United States                 402-363-6804JaeMDebord@gustr.com   
3    United States       PhanBaLiem@jourrapide.com+1 (732) 636-8246   
4    United States                  334-515-7487TimNeudorf@cuvox.de   
..             ...                                              ...   
498  United States      207-477-0579MustafaLindstrom@jourrapide.com   
499  United States               928-284-4492RumanBisliev@gustr.com   
500  United States            816-223-6007JinkedeKeizer@teleworm.us   
501  United States  ChidaluOnyekaozulu@jourrapide.com1 360 443 2060   
502  United States             PatrickGersten@rhyta.com402-848-4923   

     birthdate  weight  height   bmi  
0    7/10/1976   121.7      66  19.6  
1     4/3/1967   118.8      66  19.2  
2    2/19/1980   177.8      71  24.8  
3    7/26/1951   220.9      70  31.7  
4    2/18/1928   192.3      27  26.1  
..         ...     ...     ...   ...  
498  4/10/1959   181.1      72  24.6  
499  3/26/1948   239.6      70  34.4  
500  1/13/1971   171.2      67  26.8  
501  2/13/1952   176.9      67  27.7  
502   5/3/1954   138.2      71  19.3  

[503 rows x 14 columns]

In [208]:
treatments

given_name       surname    auralin    novodra  hba1c_start  hba1c_end  \
0     veronika      jindrová  41u - 48u          -         7.63       7.20   
1       elliot    richardson          -  40u - 45u         7.56       7.09   
2     yukitaka      takenaka          -  39u - 36u         7.68       7.25   
3         skye    gormanston  33u - 36u          -         7.97       7.62   
4       alissa        montez          -  33u - 29u         7.78       7.46   
..         ...           ...        ...        ...          ...        ...   
275     albina      zetticci  45u - 51u          -         7.93       7.73   
276       john   teichelmann          -  49u - 49u         7.90       7.58   
277     mathea       lillebø  23u - 36u          -         9.04       8.67   
278     vallie        prince  31u - 38u          -         7.64       7.28   
279     samúel  guðbrandsson  53u - 56u          -         8.00       7.64   

     hba1c_change  
0             NaN  
1            0.97  
2             NaN  
3            0.35  
4            0.32  
..            ...  
275          0.20  
276           NaN  
277          0.37  
278          0.36  
279          0.36  

[280 rows x 7 columns]

In [209]:
adverse_reactions

given_name     surname           adverse_reaction
0         berta  napolitani  injection site discomfort
1          lena        baer               hypoglycemia
2        joseph         day               hypoglycemia
3        flavia  fiorentino                      cough
4       manouck     wubbels          throat irritation
5       jasmine       sykes               hypoglycemia
6        louise     johnson               hypoglycemia
7       albinca     komavec               hypoglycemia
8           noe      aranda               hypoglycemia
9         sofia   hermansen  injection site discomfort
10        tegan     johnson                   headache
11         abel     yonatan                      cough
12    abdul-nur         isa               hypoglycemia
13         leon      scholz  injection site discomfort
14     gabriele     saenger               hypoglycemia
15       jia li        teng                     nausea
16        jakob    jakobsen               hypoglycemia
17  christopher    woodward                     nausea
18          ole    petersen               hypoglycemia
19       finley    chandler                   headache
20     anenechi       chidi               hypoglycemia
21     miłosław  wiśniewski  injection site discomfort
22        lixue       hsueh  injection site discomfort
23        merci      leroux               hypoglycemia
24         kang         mai  injection site discomfort
25       elliot  richardson               hypoglycemia
26      clinton      miller          throat irritation
27       idalia       moore               hypoglycemia
28       xiuxiu       chang               hypoglycemia
29         alex    crawford               hypoglycemia
30       monika      lončar               hypoglycemia
31       steven         roy                   headache
32      cecilie      nilsen               hypoglycemia
33    krisztina      magyar               hypoglycemia

In [210]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   patient_id    503 non-null    int64  
 1   assigned_sex  503 non-null    object 
 2   given_name    503 non-null    object 
 3   surname       503 non-null    object 
 4   address       491 non-null    object 
 5   city          491 non-null    object 
 6   state         491 non-null    object 
 7   zip_code      491 non-null    float64
 8   country       491 non-null    object 
 9   contact       491 non-null    object 
 10  birthdate     503 non-null    object 
 11  weight        503 non-null    float64
 12  height        503 non-null    int64  
 13  bmi           503 non-null    float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [211]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    280 non-null    object 
 1   surname       280 non-null    object 
 2   auralin       280 non-null    object 
 3   novodra       280 non-null    object 
 4   hba1c_start   280 non-null    float64
 5   hba1c_end     280 non-null    float64
 6   hba1c_change  171 non-null    float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [212]:
adverse_reactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   given_name        34 non-null     object
 1   surname           34 non-null     object
 2   adverse_reaction  34 non-null     object
dtypes: object(3)
memory usage: 944.0+ bytes


In [213]:
all_columns = pd.Series(list(patients) + list(treatments) + list(adverse_reactions))
all_columns[all_columns.duplicated()]

14    given_name
15       surname
21    given_name
22       surname
dtype: object

In [214]:
list(patients)

['patient_id',
 'assigned_sex',
 'given_name',
 'surname',
 'address',
 'city',
 'state',
 'zip_code',
 'country',
 'contact',
 'birthdate',
 'weight',
 'height',
 'bmi']

In [215]:
patients[patients['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

In [216]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [217]:
treatments.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

In [218]:
patients.sample(5)

patient_id assigned_sex  given_name    surname  \
192         193       female        Jade     Parker   
288         289         male      Fareeq       Kalb   
350         351         male        Diệt        Bùi   
495         496         male      Hajime    Tsukada   
66           67         male  Konstantyn  Kowalczyk   

                        address         city       state  zip_code  \
192  2957 Feathers Hooves Drive  Garden City          NY   11530.0   
288          1043 Ventura Drive   Santa Cruz  California   95060.0   
350      3788 Worthington Drive        Rylie          TX   75235.0   
495           4111 Thunder Road    San Mateo          CA   94403.0   
66          4707 Parkway Street    San Diego  California   92103.0   

           country                                      contact  birthdate  \
192  United States         631-704-6487JadeParker@superrito.com  6/19/1927   
288  United States     831-427-4114FareeqJawharKalb@armyspy.com   5/1/1985   
350  United States        BuiKimDiet@rhyta.com+1 (972) 557-0198  4/18/1936   
495  United States         650-570-4896HajimeTsukada@dayrep.com   9/5/1972   
66   United States  760-287-8295KonstantynKowalczyk@armyspy.com   3/2/1938   

     weight  height   bmi  
192   188.8      63  33.4  
288   194.5      70  27.9  
350   119.0      65  19.8  
495   168.1      66  27.1  
66    194.3      67  30.4

In [219]:
patients.surname.value_counts()

Doe            6
Jakobsen       3
Taylor         3
Ogochukwu      2
Tucker         2
              ..
Casárez        1
Mata           1
Pospíšil       1
Rukavina       1
Onyekaozulu    1
Name: surname, Length: 466, dtype: int64

In [220]:
patients.address.value_counts()

123 Main Street             6
2778 North Avenue           2
2476 Fulton Street          2
648 Old Dear Lane           2
3094 Oral Lake Road         1
                           ..
1066 Goosetown Drive        1
4291 Patton Lane            1
4643 Reeves Street          1
174 Lost Creek Road         1
3652 Boone Crockett Lane    1
Name: address, Length: 483, dtype: int64

In [221]:
patients[patients.address.duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

In [222]:
patients.weight.sort_values()

210     48.8
459    102.1
335    102.7
74     103.2
317    106.0
       ...  
144    244.9
61     244.9
283    245.5
118    254.5
485    255.9
Name: weight, Length: 503, dtype: float64

In [223]:
weight_lbs = patients[patients.surname == 'Zaitseva'].weight * 2.20462
height_in = patients[patients.surname == 'Zaitseva'].height
bmi_check = 703 * weight_lbs / (height_in * height_in)
bmi_check

210    19.055827
dtype: float64

In [224]:
patients[patients.surname == 'Zaitseva'].bmi

210    19.1
Name: bmi, dtype: float64

In [225]:
sum(treatments.auralin.isnull())

0

In [226]:
sum(treatments.novodra.isnull())

0

#### Quality
##### `patients` table
- Zip code is a float not a string
- Zip code has four digits sometimes
- Tim Neudorf height is 27 in instead of 72 in
- Full state names sometimes, abbreviations other times
- Dsvid Gustafsson
- Missing demographic information (address - contact columns) ***(can't clean)***
- Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns)
- Multiple phone number formats
- Default John Doe data
- Multiple records for Jakobsen, Gersten, Taylor
- kgs instead of lbs for Zaitseva weight

##### `treatments` table
- Missing HbA1c changes
- The letter 'u' in starting and ending doses for Auralin and Novodra
- Lowercase given names and surnames
- Missing records (280 instead of 350)
- Erroneous datatypes (auralin and novodra columns)
- Inaccurate HbA1c changes (leading 4s mistaken as 9s)
- Nulls represented as dashes (-) in auralin and novodra columns

##### `adverse_reactions` table
- Lowercase given names and surnames

#### Tidiness
- Contact column in `patients` table should be split into phone number and email
- Three variables in two columns in `treatments` table (treatment, start dose and end dose)
- Adverse reaction should be part of the `treatments` table
- Given name and surname columns in `patients` table duplicated in `treatments` and `adverse_reactions` tables

## Clean

In [227]:
patients_clean = patients.copy()
treatments_clean = treatments.copy()
adverse_reactions_clean = adverse_reactions.copy()

### Missing Data

<font color='red'>Complete the following two "Missing Data" **Define, Code, and Test** sequences after watching the *"Address Missing Data First"* video.</font>

#### `treatments`: Missing records (280 instead of 350)

##### Define
*Your definition here. Note: the missing `treatments` records are stored in a file named `treatments_cut.csv`, which you can see in this Jupyter Notebook's dashboard (click the **jupyter** logo in the top lefthand corner of this Notebook). Hint: [documentation page](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.concat.html) for the function used in the solution.*

we need to concatenate the columns from treatments_cut into treatments to fill in our missing records, we'll use concat function from pandas to do this

##### Code

In [228]:
# Your cleaning code here
treatments_cut = pd.read_csv('treatments_cut.csv')
treatments_cut.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 70 entries, 0 to 69
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    70 non-null     object 
 1   surname       70 non-null     object 
 2   auralin       70 non-null     object 
 3   novodra       70 non-null     object 
 4   hba1c_start   70 non-null     float64
 5   hba1c_end     70 non-null     float64
 6   hba1c_change  42 non-null     float64
dtypes: float64(3), object(4)
memory usage: 4.0+ KB


In [229]:
treatments_clean

given_name       surname    auralin    novodra  hba1c_start  hba1c_end  \
0     veronika      jindrová  41u - 48u          -         7.63       7.20   
1       elliot    richardson          -  40u - 45u         7.56       7.09   
2     yukitaka      takenaka          -  39u - 36u         7.68       7.25   
3         skye    gormanston  33u - 36u          -         7.97       7.62   
4       alissa        montez          -  33u - 29u         7.78       7.46   
..         ...           ...        ...        ...          ...        ...   
275     albina      zetticci  45u - 51u          -         7.93       7.73   
276       john   teichelmann          -  49u - 49u         7.90       7.58   
277     mathea       lillebø  23u - 36u          -         9.04       8.67   
278     vallie        prince  31u - 38u          -         7.64       7.28   
279     samúel  guðbrandsson  53u - 56u          -         8.00       7.64   

     hba1c_change  
0             NaN  
1            0.97  
2             NaN  
3            0.35  
4            0.32  
..            ...  
275          0.20  
276           NaN  
277          0.37  
278          0.36  
279          0.36  

[280 rows x 7 columns]

In [230]:
treatments_cut

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0       jožka   resanovič  22u - 30u          -         7.56       7.22   
1   inunnguaq    heilmann  57u - 67u          -         7.85       7.45   
2       alwin    svensson  36u - 39u          -         7.78       7.34   
3         thể       lương          -  61u - 64u         7.64       7.22   
4      amanda     ribeiro  36u - 44u          -         7.85       7.47   
..        ...         ...        ...        ...          ...        ...   
65     rovzan     kishiev  32u - 37u          -         7.75       7.41   
66      jakob    jakobsen          -  28u - 26u         7.96       7.51   
67      bernd   schneider  48u - 56u          -         7.74       7.44   
68      berta  napolitani          -  42u - 44u         7.68       7.21   
69     armina       sauvé  36u - 46u          -         7.86       7.40   

    hba1c_change  
0           0.34  
1            NaN  
2            NaN  
3           0.92  
4           0.38  
..           ...  
65          0.34  
66          0.95  
67          0.30  
68           NaN  
69           NaN  

[70 rows x 7 columns]

In [231]:
treatments_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    280 non-null    object 
 1   surname       280 non-null    object 
 2   auralin       280 non-null    object 
 3   novodra       280 non-null    object 
 4   hba1c_start   280 non-null    float64
 5   hba1c_end     280 non-null    float64
 6   hba1c_change  171 non-null    float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [232]:
treatments_clean = pd.concat([treatments_cut, treatments_clean], ignore_index='True')

##### Test

In [233]:
# Your testing code here
treatments_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 350 entries, 0 to 349
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    350 non-null    object 
 1   surname       350 non-null    object 
 2   auralin       350 non-null    object 
 3   novodra       350 non-null    object 
 4   hba1c_start   350 non-null    float64
 5   hba1c_end     350 non-null    float64
 6   hba1c_change  213 non-null    float64
dtypes: float64(3), object(4)
memory usage: 19.3+ KB


In [234]:
treatments_clean.head(20)

given_name       surname    auralin    novodra  hba1c_start  hba1c_end  \
0       jožka     resanovič  22u - 30u          -         7.56       7.22   
1   inunnguaq      heilmann  57u - 67u          -         7.85       7.45   
2       alwin      svensson  36u - 39u          -         7.78       7.34   
3         thể         lương          -  61u - 64u         7.64       7.22   
4      amanda       ribeiro  36u - 44u          -         7.85       7.47   
5        meya       sjöberg          -  42u - 42u         7.77       7.42   
6        sara         miles  30u - 35u          -         7.53       7.12   
7     frydryk       adamski  63u - 74u          -         7.75       7.27   
8        jane       citizen          -  37u - 38u         7.98       7.60   
9      angela    lavrentyev          -  28u - 24u         7.61       7.14   
10     edelma   villalpando          -  24u - 26u         7.99       7.56   
11   gabriela  michalčíková  26u - 36u          -         8.37       7.97   
12     annika         vaara          -  20u - 21u         7.73       7.34   
13   bernarda       cindrić  40u - 49u          -         7.89       7.55   
14      chiho          higa          -  46u - 46u         7.71       7.30   
15       dani        antoun  36u - 44u          -         7.73       7.34   
16     mijael        guerra  37u - 47u          -         7.85       7.46   
17       emma        barlow  48u - 57u          -         7.85       7.39   
18     flavia    fiorentino  46u - 57u          -         7.95       7.60   
19        noe        aranda  26u - 34u          -         7.51       7.17   

    hba1c_change  
0           0.34  
1            NaN  
2            NaN  
3           0.92  
4           0.38  
5           0.35  
6            NaN  
7            NaN  
8            NaN  
9           0.97  
10          0.93  
11           NaN  
12          0.39  
13          0.34  
14           NaN  
15          0.39  
16          0.39  
17           NaN  
18          0.35  
19          0.34

#### `treatments`: Missing HbA1c changes and Inaccurate HbA1c changes (leading 4s mistaken as 9s)
*Note: the "Inaccurate HbA1c changes (leading 4s mistaken as 9s)" observation, which is an accuracy issue and not a completeness issue, is included in this header because it is also fixed by the cleaning operation that fixes the missing "Missing HbA1c changes" observation. Multiple observations in one **Define, Code, and Test** header occurs multiple times in this notebook.*

##### Define
*Your definition here.*

we must redefine what variable hba1c_change is, we will do this by subtracting hba1c_start column with hba1c_end to get our wanted results

##### Code

In [235]:
# Your cleaning code here
treatments_clean.hba1c_change = treatments_clean.hba1c_start - treatments_clean.hba1c_end

##### Test

In [236]:
# Your testing code here
treatments_clean.tail(30)

given_name       surname    auralin    novodra  hba1c_start  hba1c_end  \
320        chen           yao          -  56u - 57u         7.90       7.51   
321       aksel   vestergaard          -  42u - 38u         9.62       9.29   
322       ellen         luman          -  40u - 39u         9.27       8.77   
323      albino     schiavone  35u - 43u          -         7.56       7.15   
324        jose         combs          -  39u - 36u         7.89       7.42   
325      jia li          teng  48u - 54u          -         7.66       7.32   
326       ilija        horvat  42u - 50u          -         7.77       7.38   
327    mathilde      nørgaard          -  27u - 28u         8.50       8.10   
328      csilla      herczegh          -  43u - 46u         7.71       7.27   
329     aaliyah          rice          -  31u - 31u         7.64       7.33   
330       david      beauvais          -  26u - 23u         7.87       7.47   
331    caroline        shuler          -  50u - 54u         7.63       7.27   
332        alex      crawford  51u - 62u          -         7.69       7.30   
333     rebecca      jephcott  53u - 63u          -         7.96       7.57   
334  chukwumoge     ogochukwu          -  41u - 39u         7.95       7.56   
335      fearne      mcgregor          -  27u - 29u         7.83       7.48   
336      ursula         freud  42u - 54u          -         7.75       7.46   
337        leon        scholz          -  38u - 32u         7.72       7.29   
338      yasmin        araujo          -  51u - 54u         7.82       7.36   
339      hiromu      horikawa          -  47u - 46u         7.77       7.28   
340        mika    martinsson  34u - 43u          -         7.50       7.17   
341         leo        vieira          -  30u - 33u         7.74       7.36   
342      steven           roy          -  41u - 43u         7.87       7.43   
343        kate     wilkinson  36u - 39u          -         7.72       7.20   
344        naja       enoksen  43u - 50u          -         7.98       7.59   
345      albina      zetticci  45u - 51u          -         7.93       7.73   
346        john   teichelmann          -  49u - 49u         7.90       7.58   
347      mathea       lillebø  23u - 36u          -         9.04       8.67   
348      vallie        prince  31u - 38u          -         7.64       7.28   
349      samúel  guðbrandsson  53u - 56u          -         8.00       7.64   

     hba1c_change  
320          0.39  
321          0.33  
322          0.50  
323          0.41  
324          0.47  
325          0.34  
326          0.39  
327          0.40  
328          0.44  
329          0.31  
330          0.40  
331          0.36  
332          0.39  
333          0.39  
334          0.39  
335          0.35  
336          0.29  
337          0.43  
338          0.46  
339          0.49  
340          0.33  
341          0.38  
342          0.44  
343          0.52  
344          0.39  
345          0.20  
346          0.32  
347          0.37  
348          0.36  
349          0.36

### Tidiness

<font color='red'>Complete the following four "Tidiness" **Define, Code, and Test** sequences after watching the *"Cleaning for Tidiness"* video.</font>

#### Contact column in `patients` table contains two variables: phone number and email

##### Define
*Your definition here. Hint 1: use regular expressions with pandas' [`str.extract` method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.extract.html). Here is an amazing [regex tutorial](https://regexone.com/). Hint 2: [various phone number regex patterns](https://stackoverflow.com/questions/16699007/regular-expression-to-match-standard-10-digit-phone-number). Hint 3: [email address regex pattern](http://emailregex.com/), which you might need to modify to distinguish the email from the phone number.*

we must use regex to extract information from contact column, we will create two new columns phone_number and email and extract the information from contact respectively

##### Code

In [237]:
patients_clean.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   
2               York    Nebraska   68467.0  United States   
3         Woodbridge          NJ    7095.0  United States   
4             Dothan          AL   36303.0  United States   

                                      contact  birthdate  weight  height   bmi  
0        951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66  19.6  
1       PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66  19.2  
2            402-363-6804JaeMDebord@gustr.com  2/19/1980   177.8      71  24.8  
3  PhanBaLiem@jourrapide.com+1 (732) 636-8246  7/26/1951   220.9      70  31.7  
4             334-515-7487TimNeudorf@cuvox.de  2/18/1928   192.3      27  26.1

In [238]:
# Your cleaning code here

patients_clean = patients_clean.assign(email='', phone_number='') 

patients_clean.email = patients_clean.contact.str.extract('([a-zA-Z][a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+[a-zA-Z])', expand=True)

patients_clean.phone_number = patients_clean.contact.str.extract('((?:\+\d{1,2}\s)?\(?\d{3}\)?[\s.-]?\d{3}[\s.-]?\d{4})', expand=True)

patients_clean.drop('contact', inplace=True, axis=1)

##### Test

In [239]:
# Your testing code here

patients_clean.tail()

patient_id assigned_sex given_name      surname  \
498         499         male    Mustafa    Lindström   
499         500         male      Ruman      Bisliev   
500         501       female      Jinke    de Keizer   
501         502       female    Chidalu  Onyekaozulu   
502         503         male        Pat      Gersten   

                      address           city     state  zip_code  \
498       2530 Victoria Court   Milton Mills        ME    3852.0   
499  494 Clarksburg Park Road         Sedona        AZ   86341.0   
500         649 Nutter Street  Overland Park        MO   64110.0   
501  3652 Boone Crockett Lane        Seattle        WA   98109.0   
502         2778 North Avenue           Burr  Nebraska   68324.0   

           country  birthdate  weight  height   bmi  \
498  United States  4/10/1959   181.1      72  24.6   
499  United States  3/26/1948   239.6      70  34.4   
500  United States  1/13/1971   171.2      67  26.8   
501  United States  2/13/1952   176.9      67  27.7   
502  United States   5/3/1954   138.2      71  19.3   

                                 email  phone_number  
498    MustafaLindstrom@jourrapide.com  207-477-0579  
499             RumanBisliev@gustr.com  928-284-4492  
500          JinkedeKeizer@teleworm.us  816-223-6007  
501  ChidaluOnyekaozulu@jourrapide.com  360 443 2060  
502           PatrickGersten@rhyta.com  402-848-4923

In [240]:
patients_clean.phone_number.sample(30)

12          434-509-2614
378         901 316 3780
289         504-441-7744
66          760-287-8295
125         918-714-7543
106         765 599 8847
16          559 765 7836
270    +1 (618) 512-3319
219                  NaN
306         402-780-5724
98     +1 (907) 328-4125
102    +1 (334) 261-1235
408    +1 (989) 390-0285
267    +1 (785) 823-6728
423         719-661-6624
291         302-698-2057
464         205-467-9279
410         405 449 7960
381    +1 (870) 457-5086
494    +1 (586) 790-0975
288         831-427-4114
358         203 235 1076
10          906-478-8949
379         580-622-5674
265    +1 (989) 826-7884
225         202-485-6607
59          252 291 3898
32          562-985-4582
14          918-712-3469
416         631-664-4813
Name: phone_number, dtype: object

In [241]:
sum(patients_clean.phone_number.isnull())

12

In [242]:
patients_clean.email.sort_values().head(10)

404               AaliyahRice@dayrep.com
11          Abdul-NurMummarIsa@rhyta.com
332                AbelEfrem@fleckens.hu
258              AbelYonatan@teleworm.us
305    AddolorataLombardi@jourrapide.com
118         AdibMutazzGhanem@fleckens.hu
420              AdlanShishani@gustr.com
238        AkselHVestergaard@armyspy.com
61                  AlanMilne@dayrep.com
130          AlbertRWolfe@jourrapide.com
Name: email, dtype: object

#### Three variables in two columns in `treatments` table (treatment, start dose and end dose)

##### Define
*Your definition here. Hint: use pandas' [melt function](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.melt.html) and [`str.split()` method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.split.html). Here is an excellent [`melt` tutorial](https://deparkes.co.uk/2016/10/28/reshape-pandas-data-with-melt/).*

we need to use melt on all the columns except the ones we want to pivot, then we will eliminate all the entries with '-', finally we will create new columns for start dose and end dose with split method

In [243]:
treatments_clean.head()

given_name    surname    auralin    novodra  hba1c_start  hba1c_end  \
0      jožka  resanovič  22u - 30u          -         7.56       7.22   
1  inunnguaq   heilmann  57u - 67u          -         7.85       7.45   
2      alwin   svensson  36u - 39u          -         7.78       7.34   
3        thể      lương          -  61u - 64u         7.64       7.22   
4     amanda    ribeiro  36u - 44u          -         7.85       7.47   

   hba1c_change  
0          0.34  
1          0.40  
2          0.44  
3          0.42  
4          0.38

##### Code

In [244]:
# Your cleaning code here

treatments_clean = pd.melt(treatments_clean, id_vars=['given_name', 'surname', 'hba1c_start',
                                  'hba1c_end', 'hba1c_change'], var_name = 'treatment', value_name='dose')
treatments_clean = treatments_clean[treatments_clean.dose != "-"]


In [245]:
treatments_clean.reset_index(drop=True, inplace=True)

In [246]:
treatments_clean

given_name      surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0        jožka    resanovič         7.56       7.22          0.34   auralin   
1    inunnguaq     heilmann         7.85       7.45          0.40   auralin   
2        alwin     svensson         7.78       7.34          0.44   auralin   
3       amanda      ribeiro         7.85       7.47          0.38   auralin   
4         sara        miles         7.53       7.12          0.41   auralin   
..         ...          ...          ...        ...           ...       ...   
345     yasmin       araujo         7.82       7.36          0.46   novodra   
346     hiromu     horikawa         7.77       7.28          0.49   novodra   
347        leo       vieira         7.74       7.36          0.38   novodra   
348     steven          roy         7.87       7.43          0.44   novodra   
349       john  teichelmann         7.90       7.58          0.32   novodra   

          dose  
0    22u - 30u  
1    57u - 67u  
2    36u - 39u  
3    36u - 44u  
4    30u - 35u  
..         ...  
345  51u - 54u  
346  47u - 46u  
347  30u - 33u  
348  41u - 43u  
349  49u - 49u  

[350 rows x 7 columns]

In [247]:
treatments_clean.given_name.duplicated().sum()

12

In [248]:
treatments_clean['start_dose'], treatments_clean['end_dose'] = treatments_clean['dose'].str.split(' - ', 1).str

C:\Users\Admin\AppData\Local\Temp/ipykernel_34064/1170158604.py:1: FutureWarning: Columnar iteration over characters will be deprecated in future releases.
  treatments_clean['start_dose'], treatments_clean['end_dose'] = treatments_clean['dose'].str.split(' - ', 1).str


In [249]:
treatments_clean.drop('dose', inplace=True, axis=1);

##### Test

In [250]:
# Your testing code here
treatments_clean.head(20)

given_name       surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0       jožka     resanovič         7.56       7.22          0.34   auralin   
1   inunnguaq      heilmann         7.85       7.45          0.40   auralin   
2       alwin      svensson         7.78       7.34          0.44   auralin   
3      amanda       ribeiro         7.85       7.47          0.38   auralin   
4        sara         miles         7.53       7.12          0.41   auralin   
5     frydryk       adamski         7.75       7.27          0.48   auralin   
6    gabriela  michalčíková         8.37       7.97          0.40   auralin   
7    bernarda       cindrić         7.89       7.55          0.34   auralin   
8        dani        antoun         7.73       7.34          0.39   auralin   
9      mijael        guerra         7.85       7.46          0.39   auralin   
10       emma        barlow         7.85       7.39          0.46   auralin   
11     flavia    fiorentino         7.95       7.60          0.35   auralin   
12        noe        aranda         7.51       7.17          0.34   auralin   
13   seweryna    czerwinska         7.66       7.30          0.36   auralin   
14      david     villadsen         7.86       7.40          0.46   auralin   
15        lóa  hrafnsdóttir         7.60       7.09          0.51   auralin   
16  siebrigje     koldenhof         7.90       7.59          0.31   auralin   
17     trifon      izmailov         7.68       7.28          0.40   auralin   
18      galit       casárez         7.91       7.56          0.35   auralin   
19      kifle       mustafa         7.80       7.41          0.39   auralin   

   start_dose end_dose  
0         22u      30u  
1         57u      67u  
2         36u      39u  
3         36u      44u  
4         30u      35u  
5         63u      74u  
6         26u      36u  
7         40u      49u  
8         36u      44u  
9         37u      47u  
10        48u      57u  
11        46u      57u  
12        26u      34u  
13        39u      46u  
14        26u      39u  
15        22u      31u  
16        30u      38u  
17        58u      63u  
18        27u      37u  
19        42u      50u

#### Adverse reaction should be part of the `treatments` table

In [251]:
adverse_reactions_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   given_name        34 non-null     object
 1   surname           34 non-null     object
 2   adverse_reaction  34 non-null     object
dtypes: object(3)
memory usage: 944.0+ bytes


In [252]:
treatments_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 350 entries, 0 to 349
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    350 non-null    object 
 1   surname       350 non-null    object 
 2   hba1c_start   350 non-null    float64
 3   hba1c_end     350 non-null    float64
 4   hba1c_change  350 non-null    float64
 5   treatment     350 non-null    object 
 6   start_dose    350 non-null    object 
 7   end_dose      350 non-null    object 
dtypes: float64(3), object(5)
memory usage: 22.0+ KB


##### Define
*Your definition here. Hint: [tutorial](https://chrisalbon.com/python/pandas_join_merge_dataframe.html) for the function used in the solution.*

we need to use merge on given_name so that we introduce adverse_reaction column to our treatments table

##### Code

In [253]:
treatments_clean = pd.merge(treatments_clean, adverse_reactions_clean[['given_name', 'adverse_reaction']], on='given_name', how='left')

##### Test

In [254]:
# Your testing code here
treatments_clean

given_name      surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0        jožka    resanovič         7.56       7.22          0.34   auralin   
1    inunnguaq     heilmann         7.85       7.45          0.40   auralin   
2        alwin     svensson         7.78       7.34          0.44   auralin   
3       amanda      ribeiro         7.85       7.47          0.38   auralin   
4         sara        miles         7.53       7.12          0.41   auralin   
..         ...          ...          ...        ...           ...       ...   
345     yasmin       araujo         7.82       7.36          0.46   novodra   
346     hiromu     horikawa         7.77       7.28          0.49   novodra   
347        leo       vieira         7.74       7.36          0.38   novodra   
348     steven          roy         7.87       7.43          0.44   novodra   
349       john  teichelmann         7.90       7.58          0.32   novodra   

    start_dose end_dose adverse_reaction  
0          22u      30u              NaN  
1          57u      67u              NaN  
2          36u      39u              NaN  
3          36u      44u              NaN  
4          30u      35u              NaN  
..         ...      ...              ...  
345        51u      54u              NaN  
346        47u      46u              NaN  
347        30u      33u              NaN  
348        41u      43u         headache  
349        49u      49u              NaN  

[350 rows x 9 columns]

#### Given name and surname columns in `patients` table duplicated in `treatments` and `adverse_reactions` tables  and Lowercase given names and surnames

##### Define
*Your definition here. Hint: [tutorial](https://chrisalbon.com/python/pandas_join_merge_dataframe.html) for one function used in the solution and [tutorial](http://erikrood.com/Python_References/dropping_rows_cols_pandas.html) for another function used in the solution.*

we need a subset of our patients dataframe in order to merge the duplicates in treatments, we also need patient_id from patients in treatments so that we don't need given_name, and surname in two different tables, we need to lowercase in order to merge from patients since treatments given_name and surname are lowercase, we will drop the columns after merging

##### Code

In [255]:
patients_clean

patient_id assigned_sex given_name      surname  \
0             1       female        Zoe      Wellish   
1             2       female     Pamela         Hill   
2             3         male        Jae       Debord   
3             4         male       Liêm         Phan   
4             5         male        Tim      Neudorf   
..          ...          ...        ...          ...   
498         499         male    Mustafa    Lindström   
499         500         male      Ruman      Bisliev   
500         501       female      Jinke    de Keizer   
501         502       female    Chidalu  Onyekaozulu   
502         503         male        Pat      Gersten   

                       address               city       state  zip_code  \
0         576 Brown Bear Drive  Rancho California  California   92390.0   
1    2370 University Hill Road          Armstrong    Illinois   61812.0   
2        1493 Poling Farm Road               York    Nebraska   68467.0   
3          2335 Webster Street         Woodbridge          NJ    7095.0   
4         1428 Turkey Pen Lane             Dothan          AL   36303.0   
..                         ...                ...         ...       ...   
498        2530 Victoria Court       Milton Mills          ME    3852.0   
499   494 Clarksburg Park Road             Sedona          AZ   86341.0   
500          649 Nutter Street      Overland Park          MO   64110.0   
501   3652 Boone Crockett Lane            Seattle          WA   98109.0   
502          2778 North Avenue               Burr    Nebraska   68324.0   

           country  birthdate  weight  height   bmi  \
0    United States  7/10/1976   121.7      66  19.6   
1    United States   4/3/1967   118.8      66  19.2   
2    United States  2/19/1980   177.8      71  24.8   
3    United States  7/26/1951   220.9      70  31.7   
4    United States  2/18/1928   192.3      27  26.1   
..             ...        ...     ...     ...   ...   
498  United States  4/10/1959   181.1      72  24.6   
499  United States  3/26/1948   239.6      70  34.4   
500  United States  1/13/1971   171.2      67  26.8   
501  United States  2/13/1952   176.9      67  27.7   
502  United States   5/3/1954   138.2      71  19.3   

                                 email       phone_number  
0             ZoeWellish@superrito.com       951-719-9170  
1                 PamelaSHill@cuvox.de  +1 (217) 569-3204  
2                 JaeMDebord@gustr.com       402-363-6804  
3            PhanBaLiem@jourrapide.com  +1 (732) 636-8246  
4                  TimNeudorf@cuvox.de       334-515-7487  
..                                 ...                ...  
498    MustafaLindstrom@jourrapide.com       207-477-0579  
499             RumanBisliev@gustr.com       928-284-4492  
500          JinkedeKeizer@teleworm.us       816-223-6007  
501  ChidaluOnyekaozulu@jourrapide.com       360 443 2060  
502           PatrickGersten@rhyta.com       402-848-4923  

[503 rows x 15 columns]

In [256]:
# Your cleaning code here
sub_df = patients_clean[['patient_id', 'given_name', 'surname']]
sub_df.given_name = sub_df.given_name.str.lower()
sub_df.surname = sub_df.surname.str.lower()

treatments_clean = pd.merge(treatments_clean, sub_df, on=['given_name', 'surname'])
treatments_clean = treatments_clean.drop(['given_name', 'surname'], axis=1)

C:\Users\Admin\anaconda3\envs\Udacity\lib\site-packages\pandas\core\generic.py:5516: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[name] = value


##### Test

In [257]:
# Your testing code here
treatments_clean

hba1c_start  hba1c_end  hba1c_change treatment start_dose end_dose  \
0           7.56       7.22          0.34   auralin        22u      30u   
1           7.85       7.45          0.40   auralin        57u      67u   
2           7.78       7.34          0.44   auralin        36u      39u   
3           7.85       7.47          0.38   auralin        36u      44u   
4           7.53       7.12          0.41   auralin        30u      35u   
..           ...        ...           ...       ...        ...      ...   
344         7.82       7.36          0.46   novodra        51u      54u   
345         7.77       7.28          0.49   novodra        47u      46u   
346         7.74       7.36          0.38   novodra        30u      33u   
347         7.87       7.43          0.44   novodra        41u      43u   
348         7.90       7.58          0.32   novodra        49u      49u   

    adverse_reaction  patient_id  
0                NaN         468  
1                NaN         107  
2                NaN         438  
3                NaN         144  
4                NaN         260  
..               ...         ...  
344              NaN          59  
345              NaN         440  
346              NaN         434  
347         headache         394  
348              NaN         141  

[349 rows x 8 columns]

### Quality

<font color='red'>Complete the remaining "Quality" **Define, Code, and Test** sequences after watching the *"Cleaning for Quality"* video.</font>

#### Zip code is a float not a string and Zip code has four digits sometimes

##### Define
*Your definition here. Hint: see the "Data Cleaning Process" page.*

we must convert the floats to strings then we need to replace the . in our zipcodes, finally some of our zipcodes only have 4 digits so we need to add 0.

In [258]:
patients_clean.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state  zip_code        country  birthdate  weight  \
0  Rancho California  California   92390.0  United States  7/10/1976   121.7   
1          Armstrong    Illinois   61812.0  United States   4/3/1967   118.8   
2               York    Nebraska   68467.0  United States  2/19/1980   177.8   
3         Woodbridge          NJ    7095.0  United States  7/26/1951   220.9   
4             Dothan          AL   36303.0  United States  2/18/1928   192.3   

   height   bmi                      email       phone_number  
0      66  19.6   ZoeWellish@superrito.com       951-719-9170  
1      66  19.2       PamelaSHill@cuvox.de  +1 (217) 569-3204  
2      71  24.8       JaeMDebord@gustr.com       402-363-6804  
3      70  31.7  PhanBaLiem@jourrapide.com  +1 (732) 636-8246  
4      27  26.1        TimNeudorf@cuvox.de       334-515-7487

##### Code

In [259]:
patients_clean.zip_code = patients_clean.zip_code.astype(str).str.replace('.', '').str[:-1].str.zfill(5)

patients_clean.zip_code = patients_clean.zip_code.replace('000na', np.nan)

C:\Users\Admin\AppData\Local\Temp/ipykernel_34064/3321460099.py:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  patients_clean.zip_code = patients_clean.zip_code.astype(str).str.replace('.', '').str[:-1].str.zfill(5)


##### Test

In [260]:
# Your testing code here
patients_clean

patient_id assigned_sex given_name      surname  \
0             1       female        Zoe      Wellish   
1             2       female     Pamela         Hill   
2             3         male        Jae       Debord   
3             4         male       Liêm         Phan   
4             5         male        Tim      Neudorf   
..          ...          ...        ...          ...   
498         499         male    Mustafa    Lindström   
499         500         male      Ruman      Bisliev   
500         501       female      Jinke    de Keizer   
501         502       female    Chidalu  Onyekaozulu   
502         503         male        Pat      Gersten   

                       address               city       state zip_code  \
0         576 Brown Bear Drive  Rancho California  California    92390   
1    2370 University Hill Road          Armstrong    Illinois    61812   
2        1493 Poling Farm Road               York    Nebraska    68467   
3          2335 Webster Street         Woodbridge          NJ    07095   
4         1428 Turkey Pen Lane             Dothan          AL    36303   
..                         ...                ...         ...      ...   
498        2530 Victoria Court       Milton Mills          ME    03852   
499   494 Clarksburg Park Road             Sedona          AZ    86341   
500          649 Nutter Street      Overland Park          MO    64110   
501   3652 Boone Crockett Lane            Seattle          WA    98109   
502          2778 North Avenue               Burr    Nebraska    68324   

           country  birthdate  weight  height   bmi  \
0    United States  7/10/1976   121.7      66  19.6   
1    United States   4/3/1967   118.8      66  19.2   
2    United States  2/19/1980   177.8      71  24.8   
3    United States  7/26/1951   220.9      70  31.7   
4    United States  2/18/1928   192.3      27  26.1   
..             ...        ...     ...     ...   ...   
498  United States  4/10/1959   181.1      72  24.6   
499  United States  3/26/1948   239.6      70  34.4   
500  United States  1/13/1971   171.2      67  26.8   
501  United States  2/13/1952   176.9      67  27.7   
502  United States   5/3/1954   138.2      71  19.3   

                                 email       phone_number  
0             ZoeWellish@superrito.com       951-719-9170  
1                 PamelaSHill@cuvox.de  +1 (217) 569-3204  
2                 JaeMDebord@gustr.com       402-363-6804  
3            PhanBaLiem@jourrapide.com  +1 (732) 636-8246  
4                  TimNeudorf@cuvox.de       334-515-7487  
..                                 ...                ...  
498    MustafaLindstrom@jourrapide.com       207-477-0579  
499             RumanBisliev@gustr.com       928-284-4492  
500          JinkedeKeizer@teleworm.us       816-223-6007  
501  ChidaluOnyekaozulu@jourrapide.com       360 443 2060  
502           PatrickGersten@rhyta.com       402-848-4923  

[503 rows x 15 columns]

In [261]:
patients_clean.zip_code.value_counts()

12345    6
30303    4
10004    4
11530    3
35203    3
        ..
66061    1
29180    1
59522    1
85023    1
64110    1
Name: zip_code, Length: 430, dtype: int64

In [262]:
patients_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 15 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   patient_id    503 non-null    int64  
 1   assigned_sex  503 non-null    object 
 2   given_name    503 non-null    object 
 3   surname       503 non-null    object 
 4   address       491 non-null    object 
 5   city          491 non-null    object 
 6   state         491 non-null    object 
 7   zip_code      491 non-null    object 
 8   country       491 non-null    object 
 9   birthdate     503 non-null    object 
 10  weight        503 non-null    float64
 11  height        503 non-null    int64  
 12  bmi           503 non-null    float64
 13  email         491 non-null    object 
 14  phone_number  491 non-null    object 
dtypes: float64(2), int64(2), object(11)
memory usage: 59.1+ KB


#### Tim Neudorf height is 27 in instead of 72 in

##### Define
*Your definition here.*

simply replace 27 with 72

##### Code

In [263]:
# Your cleaning code here

patients_clean.loc[patients_clean.height==27, 'height'] = 72

##### Test

In [264]:
# Your testing code here
patients_clean[patients_clean.height == 27]

Empty DataFrame
Columns: [patient_id, assigned_sex, given_name, surname, address, city, state, zip_code, country, birthdate, weight, height, bmi, email, phone_number]
Index: []

#### Full state names sometimes, abbreviations other times

##### Define
*Your definition here. Hint: [tutorial](https://chrisalbon.com/python/pandas_apply_operations_to_dataframes.html) for method used in solution.*

need to convert all the full state names into abbreviations to have consistent state column

##### Code

In [265]:
# Your cleaning code here

patients_clean.state.value_counts()

California    36
TX            32
New York      25
CA            24
NY            22
MA            22
PA            18
GA            15
Illinois      14
OH            14
OK            13
MI            13
Florida       13
LA            13
NJ            12
VA            11
MS            10
WI            10
IL            10
IN             9
MN             9
FL             9
AL             9
TN             9
WA             8
NC             8
KY             8
MO             7
ID             6
NV             6
KS             6
SC             5
IA             5
CT             5
ME             4
RI             4
Nebraska       4
ND             4
CO             4
AZ             4
AR             4
MD             3
DE             3
WV             3
SD             3
OR             3
NE             2
MT             2
VT             2
DC             2
AK             1
WY             1
NH             1
NM             1
Name: state, dtype: int64

In [266]:
state_dict = {'California' : 'CA', 'New York' : 'NY', 'Illinois' : 'IL', 'Florida' : 'FL', 'Nebraska' : 'NE'}

def abbreviate_state(df):

    if df['state'] in state_dict.keys():

        abbrev = state_dict[df['state']]

        return abbrev

    else:

        return df['state']

patients_clean['state'] = patients_clean.apply(abbreviate_state, axis=1)

        
        

##### Test

In [267]:
# Your testing code here
patients_clean.state

0      CA
1      IL
2      NE
3      NJ
4      AL
       ..
498    ME
499    AZ
500    MO
501    WA
502    NE
Name: state, Length: 503, dtype: object

#### Dsvid Gustafsson

##### Define
*Your definition here.*

we just need to rename it using replace

##### Code

In [268]:
# Your cleaning code here
patients_clean.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city state zip_code        country  birthdate  weight  height  \
0  Rancho California    CA    92390  United States  7/10/1976   121.7      66   
1          Armstrong    IL    61812  United States   4/3/1967   118.8      66   
2               York    NE    68467  United States  2/19/1980   177.8      71   
3         Woodbridge    NJ    07095  United States  7/26/1951   220.9      70   
4             Dothan    AL    36303  United States  2/18/1928   192.3      72   

    bmi                      email       phone_number  
0  19.6   ZoeWellish@superrito.com       951-719-9170  
1  19.2       PamelaSHill@cuvox.de  +1 (217) 569-3204  
2  24.8       JaeMDebord@gustr.com       402-363-6804  
3  31.7  PhanBaLiem@jourrapide.com  +1 (732) 636-8246  
4  26.1        TimNeudorf@cuvox.de       334-515-7487

In [269]:
patients_clean.given_name = patients_clean.given_name.replace('Dsvid', 'David')

##### Test

In [270]:
# Your testing code here
patients_clean[patients_clean['surname'] == 'Gustafsson']

patient_id assigned_sex given_name     surname             address  \
8           9         male      David  Gustafsson  1790 Nutter Street   

          city state zip_code        country birthdate  weight  height   bmi  \
8  Kansas City    MO    64105  United States  3/6/1937   163.9      66  26.5   

                         email  phone_number  
8  DavidGustafsson@armyspy.com  816-265-9578

#### Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns) and Erroneous datatypes (auralin and novodra columns) and The letter 'u' in starting and ending doses for Auralin and Novodra

##### Define
*Your definition here. Hint: [documentation page](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.astype.html) for one method used in solution, [documentation page](http://pandas.pydata.org/pandas-docs/version/0.20/generated/pandas.to_datetime.html) for one function used in the solution, and [documentation page](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.strip.html) for another method used in the solution.*

we need to change assigned sex and state to categorical datatypes using astype, then we will use to_datetime to change birthdate to its proper format, and finally we will use str.strip to strip the 'u' from our dosages


##### Code

In [271]:
# Your cleaning code here

patients_clean.assigned_sex = patients_clean.assigned_sex.astype('category')
patients_clean.state = patients_clean.state.astype('category')
patients_clean.birthdate = pd.to_datetime(patients_clean.birthdate)

In [272]:
treatments_clean.head()

hba1c_start  hba1c_end  hba1c_change treatment start_dose end_dose  \
0         7.56       7.22          0.34   auralin        22u      30u   
1         7.85       7.45          0.40   auralin        57u      67u   
2         7.78       7.34          0.44   auralin        36u      39u   
3         7.85       7.47          0.38   auralin        36u      44u   
4         7.53       7.12          0.41   auralin        30u      35u   

  adverse_reaction  patient_id  
0              NaN         468  
1              NaN         107  
2              NaN         438  
3              NaN         144  
4              NaN         260

In [273]:
treatments_clean.start_dose = treatments_clean.start_dose.str.rstrip('u').astype(int)
treatments_clean.end_dose = treatments_clean.end_dose.str.rstrip('u').astype(int)

##### Test

In [274]:
# Your testing code here
patients_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 15 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   patient_id    503 non-null    int64         
 1   assigned_sex  503 non-null    category      
 2   given_name    503 non-null    object        
 3   surname       503 non-null    object        
 4   address       491 non-null    object        
 5   city          491 non-null    object        
 6   state         491 non-null    category      
 7   zip_code      491 non-null    object        
 8   country       491 non-null    object        
 9   birthdate     503 non-null    datetime64[ns]
 10  weight        503 non-null    float64       
 11  height        503 non-null    int64         
 12  bmi           503 non-null    float64       
 13  email         491 non-null    object        
 14  phone_number  491 non-null    object        
dtypes: category(2), datetime64[ns](1), float

In [275]:
treatments_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 349 entries, 0 to 348
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   hba1c_start       349 non-null    float64
 1   hba1c_end         349 non-null    float64
 2   hba1c_change      349 non-null    float64
 3   treatment         349 non-null    object 
 4   start_dose        349 non-null    int32  
 5   end_dose          349 non-null    int32  
 6   adverse_reaction  37 non-null     object 
 7   patient_id        349 non-null    int64  
dtypes: float64(3), int32(2), int64(1), object(2)
memory usage: 21.8+ KB


In [276]:
treatments_clean.start_dose.value_counts()

42    20
36    18
37    17
41    17
39    15
31    15
30    14
40    14
43    13
26    12
48    12
35    12
32    11
29    11
44    11
51    10
38    10
34     9
55     8
27     8
49     8
47     8
33     7
28     6
54     6
24     6
46     6
53     6
22     5
25     4
50     4
45     3
57     3
58     3
52     3
23     3
61     2
63     2
56     2
62     1
20     1
19     1
60     1
59     1
Name: start_dose, dtype: int64

In [277]:
treatments_clean.end_dose.value_counts()

39    20
36    18
38    16
42    15
43    13
50    13
46    13
37    13
32    12
41    12
45    12
47    11
54    10
40    10
44    10
31     9
30     8
33     8
49     8
48     8
35     7
57     7
29     7
58     7
55     6
51     6
23     6
60     6
56     6
53     5
34     5
27     5
59     4
24     4
68     3
52     3
64     3
67     3
61     2
62     2
26     2
21     2
63     2
25     2
66     1
75     1
69     1
74     1
28     1
Name: end_dose, dtype: int64

#### multiple phone number formats

##### Define
*Your definition here. Hint: helpful [Stack Overflow answer](https://stackoverflow.com/a/123681).*

we need to remove any +, (, -, we will use replace with a regex to replace with empty string then pad to 11 digits

##### Code

In [278]:
# Your cleaning code here
patients_clean.phone_number.value_counts()

1234567890           6
402-848-4923         2
304-438-2648         2
+1 (845) 858-7707    2
952 431 5166         1
                    ..
+1 (828) 586-5050    1
+1 (918) 971-5864    1
920-849-0384         1
+1 (612) 208-2965    1
360 443 2060         1
Name: phone_number, Length: 483, dtype: int64

In [279]:
patients_clean.phone_number = patients_clean.phone_number.str.replace(r'\D+', '').str.pad(11, fillchar='1')

C:\Users\Admin\AppData\Local\Temp/ipykernel_34064/3922059896.py:1: FutureWarning: The default value of regex will change from True to False in a future version.
  patients_clean.phone_number = patients_clean.phone_number.str.replace(r'\D+', '').str.pad(11, fillchar='1')


##### Test

In [280]:
# Your testing code here
patients_clean.phone_number.value_counts()

11234567890    6
14028484923    2
13044382648    2
18458587707    2
19524315166    1
              ..
18285865050    1
19189715864    1
19208490384    1
16122082965    1
13604432060    1
Name: phone_number, Length: 483, dtype: int64

#### Default John Doe data

##### Define
*Your definition here. Recall that it is assumed that the data that this John Doe data displaced is not recoverable.*

we need to remove all instaces of John Doe since the data is not recoverable.

In [281]:
patients_clean.query('given_name == "John" & surname == "Doe"')

patient_id assigned_sex given_name surname          address      city  \
215         216         male       John     Doe  123 Main Street  New York   
229         230         male       John     Doe  123 Main Street  New York   
237         238         male       John     Doe  123 Main Street  New York   
244         245         male       John     Doe  123 Main Street  New York   
251         252         male       John     Doe  123 Main Street  New York   
277         278         male       John     Doe  123 Main Street  New York   

    state zip_code        country  birthdate  weight  height   bmi  \
215    NY    12345  United States 1975-01-01   180.0      72  24.4   
229    NY    12345  United States 1975-01-01   180.0      72  24.4   
237    NY    12345  United States 1975-01-01   180.0      72  24.4   
244    NY    12345  United States 1975-01-01   180.0      72  24.4   
251    NY    12345  United States 1975-01-01   180.0      72  24.4   
277    NY    12345  United States 1975-01-01   180.0      72  24.4   

                 email phone_number  
215  johndoe@email.com  11234567890  
229  johndoe@email.com  11234567890  
237  johndoe@email.com  11234567890  
244  johndoe@email.com  11234567890  
251  johndoe@email.com  11234567890  
277  johndoe@email.com  11234567890

##### Code

In [282]:
# Your cleaning code here

patients_clean.drop(patients_clean.loc[patients_clean['surname'] == 'Doe'].index, inplace=True)

##### Test

In [283]:
# Your testing code here
patients_clean.given_name.value_counts()

John      3
David     3
Robert    3
Søren     3
Abel      2
         ..
Markus    1
Galit     1
Melisa    1
Peter     1
Pat       1
Name: given_name, Length: 469, dtype: int64

In [284]:
patients_clean.address.value_counts()

2778 North Avenue           2
2476 Fulton Street          2
648 Old Dear Lane           2
576 Brown Bear Drive        1
2272 Williams Avenue        1
                           ..
1066 Goosetown Drive        1
4291 Patton Lane            1
4643 Reeves Street          1
174 Lost Creek Road         1
3652 Boone Crockett Lane    1
Name: address, Length: 482, dtype: int64

#### Multiple records for Jakobsen, Gersten, Taylor

##### Define
*Your definition here.*

we need to remove the nicknamed entries using queries and drop methods

##### Code

In [285]:
patients_clean = patients_clean[~((patients_clean.address.duplicated()) & patients_clean.address.notnull())]

##### Test

In [286]:
# Your testing code here
patients_clean.query('surname == "Jakobsen"')

patient_id assigned_sex given_name   surname             address  \
24           25         male      Jakob  Jakobsen   648 Old Dear Lane   
432         433       female      Karen  Jakobsen  1690 Fannie Street   

            city state zip_code        country  birthdate  weight  height  \
24   Port Jervis    NY    12771  United States 1985-08-01   155.8      67   
432      Houston    TX    77020  United States 1962-11-25   185.2      67   

      bmi                         email phone_number  
24   24.4     JakobCJakobsen@einrot.com  18458587707  
432  29.0  KarenJakobsen@jourrapide.com  19792030438

In [287]:
patients_clean.query('surname == "Gersten"')

patient_id assigned_sex given_name  surname            address  city  \
97          98         male    Patrick  Gersten  2778 North Avenue  Burr   

   state zip_code        country  birthdate  weight  height   bmi  \
97    NE    68324  United States 1954-05-03   138.2      71  19.3   

                       email phone_number  
97  PatrickGersten@rhyta.com  14028484923

In [288]:
patients_clean.query('surname == "Taylor"')

patient_id assigned_sex given_name surname             address      city  \
131         132       female     Sandra  Taylor  2476 Fulton Street  Rainelle   
426         427         male    Rogelio  Taylor  4064 Marigold Lane     Miami   

    state zip_code        country  birthdate  weight  height   bmi  \
131    WV    25962  United States 1960-10-23   206.1      64  35.4   
426    FL    33179  United States 1992-09-02   186.6      69  27.6   

                          email phone_number  
131    SandraCTaylor@dayrep.com  13044382648  
426  RogelioJTaylor@teleworm.us  13054346299

#### kgs instead of lbs for Zaitseva weight

##### Define
*Your definition here.*

we simply convert the kgs to lbs for zaitseva's weight

##### Code

In [289]:
# Your cleaning code here

patients_clean.query('surname == "Zaitseva"')

patient_id assigned_sex given_name   surname              address  \
210         211       female    Camilla  Zaitseva  4689 Briarhill Lane   

        city state zip_code        country  birthdate  weight  height   bmi  \
210  Wooster    OH    44691  United States 1938-11-26    48.8      63  19.1   

                             email phone_number  
210  CamillaZaitseva@superrito.com  13302022145

In [290]:
kg = patients_clean[patients_clean['surname'] == 'Zaitseva']['weight'].values
patients_clean.loc[patients_clean.surname == 'Zaitseva','weight'] = kg * 2.20462

C:\Users\Admin\anaconda3\envs\Udacity\lib\site-packages\pandas\core\indexing.py:1773: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_column(ilocs[0], value, pi)


##### Test

In [291]:
# Your testing code here
patients_clean.weight.sort_values()

459    102.1
335    102.7
74     103.2
317    106.0
171    106.5
       ...  
144    244.9
61     244.9
283    245.5
118    254.5
485    255.9
Name: weight, Length: 494, dtype: float64

In [292]:
patients_clean.query('surname == "Zaitseva"')

patient_id assigned_sex given_name   surname              address  \
210         211       female    Camilla  Zaitseva  4689 Briarhill Lane   

        city state zip_code        country  birthdate      weight  height  \
210  Wooster    OH    44691  United States 1938-11-26  107.585456      63   

      bmi                          email phone_number  
210  19.1  CamillaZaitseva@superrito.com  13302022145